docs/async/async-csharp-vb.rst

```diff
@@ -0,0 +1,200 @@
+Async Programming in C#/VB
+==========================
+By `Phillip Carter`_
+
+Async programming is traditionally associated with callbacks and a general lack of straightforward code.  One might even say that straightforward code and scalable code are often at odds with one another.  This is not a problem with the C#/VB async model!  It is a language-level construct which hides complexity and allows you to focus on the problem at hand without making the code confusing in the process.
+
+The core of this are the ``async`` and ``await`` keywords (``Async`` and ``Await`` in VB), which allow for simple changes to code with a huge impact.  For example:
+
+.. code-block:: c#
+
+	button.Clicked += async (o, e) =>
+	{
+	    var data = await _httpClient.DownloadStringAsync(URL);
+	    DisplayData(data);
+	};
+	
+And that's it!  No callbacks and no library to wrestle with.
+
+For those who are more theoretically-inclined, the model used by C# and VB is an implementation of the `Future and Promise concurrency model <https://en.wikipedia.org/wiki/Futures_and_promises>`_.
+
+A few important things to know before continuing:
+
+* Async code relies on ``Task<T>`` and ``Task``, which are constructs used to abstract the notion of a job that you need to wait on for a result.  They also act as the return types, except for the case of Event Handlers (these return ``void``).  `More on Task and Task<T>`_.
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` the completion of a task.   Application of ``await`` is optional, but recommended in most cases.
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the calling method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
+* Unless an async method has an ``await`` inside its body, it will never yield!
+* Although not strictly required, ``async void`` should **only** be used on Event Handlers.  More on this later.
+
+Bigger Example (C#)
+------------------
+
+The following example shows somewhat of a "real life" scenario: getting directions to a user for a ridesharing app.
+
+Client app snippet:
+
+.. code-block:: c#
+
+	private async void GetDirectionsForPickup_Pressed(object sender, CustomUserDataEventArgs e)
+	{
+	    var user = e.Source as BasicUserData;
+
+	    // _directionsClient is a made-up object that would encapsulate certain complexity,
+	    // such as getting the location from the device.
+	    var getDirectionsTask = _directionsClient.GetDrivingDirectionsToLocation(user.LastKnownLocation);
+		
+	    // This independent work can be done concurrently
+	    // since it doesn't rely on directions!
+	    NotifyUserOfPickupAsync(user.Contact);
```

Comment:
This line doesn't seem right. I would expect you want to assign this to a variable of Task<T> and then await it later. Otherwise, it won't be concurrent. I also don't expect that this code will compile as-is.